compaction/compaction.cc

```diff
@@ -1291,6 +1291,11 @@ class cleanup_compaction final : public regular_compaction {
     // Called in a seastar thread
     dht::partition_range_vector
     get_ranges_for_invalidation(const std::vector<shared_sstable>& sstables) {
+        if (!_owned_ranges) {
```

Comment:
> owned_ranges_ptr is set to null because
cleanup is not required for tables using tablets. As a result, cache
invalidation is also not required for any ranges. 

I don't understand this. Why is cleanup not cache invalidation not required?

Cleanup is required, it just end being a no-op. We remove all the sstables containing the owned ranges being lost, and so a cleanup would find nothing to do. But the data is still in cache and has to be removed.

Reply:
I don't remember why we picked cleanup_compaction class for upgrade. Migration cleanup happens when tablet range is still owned, so we don't use cleanup_compaction for it. One way to make it work is to feed an empty owned range if we detect underlying tablet is at cleanup stage.

@bhalevy reviewing this area, I think we have a potential data resurrection after we now piggyback on other compaction types for cleanup work. we're not invalidating cache on them, so regain of ownership could cause resurrection. With tablets, it's not a problem as we explicitly invalidate the cache during cleanup step, but with vnodes, it is.

Reply:
Since commit f1bbf705f9b6faa406baac04f418a93b09a53c19, upgrade and other types might perform cleanup work optionally, if owned_ranges is engaged. Not to confuse desingaged owned_ranges with empty (meaning everything can be cleaned up). So I think the patch is right as is, to fix #17452. #17452 doesn't happen in the context of tablet migration, but rather upgrade compaction which reuses cleanup_compaction class that tries to invalidate unconditionally. And since commit 7a98877798817, owned_ranges is desingaged with tablets. Alternatively, we supply empty owned_ranges if underlying tablet is at cleanup stage, otherwise the tablet range itself. But currently we don't rely on compaction layer at all for tablet cleanup work, it's just plain deletion (after flushing memtable and stopping ongoing compaction) and invalidation.